src/internal/attr.c

```diff
@@ -85,12 +91,16 @@ r_obj* ffi_set_names(r_obj* x, r_obj* mold, r_obj* nm, r_obj* env) {
 
     nm = KEEP_N(rlang_as_function(nm, env), &n_kept);
     nm = KEEP_N(eval_fn_dots(nm, mold, dots, env), &n_kept);
-  } else {
-    if (r_length(dots) > 0) {
-      nm = KEEP_N(eval_fn_dots(c_fn, nm, dots, env), &n_kept);
-    }
+  } else if (r_length(dots) > 0) {
+    nm = KEEP_N(eval_fn_dots(c_fn, nm, dots, env), &n_kept);
+  }
 
+  if (r_typeof(nm) != R_TYPE_character || r_is_object(nm)) {
     nm = KEEP_N(eval_as_character(nm, env), &n_kept);
+
+    if (r_typeof(nm) != R_TYPE_character) {
+      r_abort("`nm` must be `NULL` or a character vector.");
+    }
   }
```

Comment:
I tweaked this section a little

- The `nm == function` case didn't call `as.character()` on the result of the function, but I felt like it was reasonable to do so since we call `as.character()` on the result of `set_names(x, 1, 2)` when `...` are combined with `c()`, and on `set_names(x, 1)` when `nm` just isn't a character vector.

- We also had some redundant "if not character, then error" checks below that I've removed in favor of just this single character check.